File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.java
import java.util.*;

class Solution {
    public int[] solution(String[] genres, int[] plays) {
        List<Integer> answer = new ArrayList<>();
        // 장르별 Map 부터
        Map<String, Integer> genreMap = new HashMap<>();
        for(int i=0; i<genres.length; i++){
            String curGen = genres[i];
            int curCnt = plays[i];
            
            genreMap.put(curGen, genreMap.getOrDefault(curGen, 0)+curCnt);
        }
        
        List<String> genreRank = new ArrayList<>(genreMap.keySet());
        genreRank.sort((e1, e2) -> Integer.compare(genreMap.get(e2), genreMap.get(e1)));

        int idx = 0;
        for(String genre : genreRank){
            List<Integer> list = new ArrayList<>();
            for(int i=0; i<genres.length; i++){
                if(genres[i].equals(genre)){
                    list.add(i);
                }
            }
            list.sort((e1, e2) -> Integer.compare(plays[e2], plays[e1]));
            int cnt = 0;
            for(int num : list){
                if(cnt >= 2) break;
                answer.add(num);
                cnt++;
            }
        }
        int[] ans = new int[answer.size()];
        for(int i=0; i<answer.size(); i++){
            ans[i] = answer.get(i);
        }
        return ans;
    }
}

File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.cs
using System;
using System.Collections.Generic;
using System.Linq;

public class Solution
{
    public int[] solution(string[] genres, int[] plays)
    {
        List<int> answer = new List<int>();

        Dictionary<string, int> genreMap = new Dictionary<string, int>();

        for (int i = 0; i < genres.Length; i++)
        {
            string curGen = genres[i];
            int curCnt = plays[i];

            if (genreMap.ContainsKey(curGen))
                genreMap[curGen] += curCnt;
            else
                genreMap[curGen] = curCnt;
        }

        List<string> genreRank = genreMap.Keys.ToList();

        genreRank.Sort((e1, e2) =>
            genreMap[e2].CompareTo(genreMap[e1])
        );

        foreach (string genre in genreRank)
        {
            List<int> list = new List<int>();

            for (int i = 0; i < genres.Length; i++)
            {
                if (genres[i] == genre)
                {
                    list.Add(i);
                }
            }

            list.Sort((e1, e2) =>
                plays[e2].CompareTo(plays[e1])
            );

            int cnt = 0;

            foreach (int num in list)
            {
                if (cnt >= 2)
                    break;

                answer.Add(num);
                cnt++;
            }
        }

        return answer.ToArray();
    }
}